compiler/pipes/register-kphp-configuration.cpp

```diff
@@ -31,43 +31,110 @@ void RegisterKphpConfiguration::on_start() {
+  parse_palette(c.value, palette);
+  G->set_function_palette(std::move(palette));
+}
+
+void RegisterKphpConfiguration::parse_palette(VertexPtr const_val, function_palette::Palette &palette) {
+  auto arr = const_val.try_as<op_array>();
+  kphp_error_return(arr, fmt_format("{}::{} must be a constexpr array",
+                                    configuration_class_name_, function_color_palette_name_));
+
+  for (const auto &opt : arr->args()) {
+    const auto ruleset_arr = opt.try_as<op_array>();
+    kphp_error_act(ruleset_arr, fmt_format("{}::{} must contain constexpr arrays (rulesets)",
+                                           configuration_class_name_, function_color_palette_name_), continue);
+
+    parse_palette_ruleset(ruleset_arr, palette);
+  }
+}
+
+void RegisterKphpConfiguration::parse_palette_ruleset(VertexAdaptor<op_array> arr, function_palette::Palette &palette) {
+  function_palette::PaletteRuleset ruleset;
+
+  for (const auto &rule_node : arr->args()) {
+    auto pair = rule_node.try_as<op_double_arrow>();
+    kphp_error_act(pair, fmt_format("{}::{} rule must be an associative map",
+                                    configuration_class_name_, function_color_palette_name_), continue);
+
+    parse_palette_rule(pair, palette, ruleset);
+  }
+
+  palette.add_ruleset(std::move(ruleset));
+}
+
+void RegisterKphpConfiguration::parse_palette_rule(VertexAdaptor<op_double_arrow> pair, function_palette::Palette &palette, function_palette::PaletteRuleset &add_to) {
+  auto parse_rule_key_colors = [this, &palette](VertexPtr key) -> std::vector<function_palette::color_t> {
+    const auto *rule_string = GenTree::get_constexpr_string(key);
+    auto color_names_strings = split(rule_string != nullptr ? *rule_string : "", ' ');
+    kphp_error(!color_names_strings.empty(), fmt_format("{}::{} map keys must be non-empty constexpr strings",
+                                                        configuration_class_name_, function_color_palette_name_));
+    auto mapper = vk::make_transform_iterator_range([&palette](auto color_name) {
+      return palette.register_color_name(color_name);
+    }, color_names_strings.begin(), color_names_strings.end());
+    return {mapper.begin(), mapper.end()};
+  };
+
+  auto parse_rule_value_error = [this](VertexPtr value) -> std::string {
+    const std::string *errtext_val = GenTree::get_constexpr_string(value);
+    const auto ok_val = value.try_as<op_int_const>();
+    kphp_error(errtext_val || (ok_val && ok_val->get_string() == "1"), fmt_format("{}::{} map values must be constexpr strings or number 1",
+                                                                                  configuration_class_name_, function_color_palette_name_));
+    return errtext_val != nullptr ? *errtext_val : "";
+  };
+
+  auto colors = parse_rule_key_colors(pair->key());
+  auto error = parse_rule_value_error(pair->value());
+  add_to.add_rule(function_palette::PaletteRule(std::move(colors), std::move(error)));
```

Comment:
```suggestion
  add_to.add_rule(function_palette::PaletteRule{std::move(colors), std::move(error)});
```
I think, it's a bit more readable, because it highlights object creation among function calls

Reply:
Done.